ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/HealthCheckResource.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.resources;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.confluent.ksql.rest.entity.HealthCheckResponse;
+import io.confluent.ksql.rest.entity.Versions;
+import io.confluent.ksql.rest.healthcheck.HealthCheckAgent;
+import io.confluent.ksql.rest.server.KsqlRestConfig;
+import io.confluent.ksql.rest.server.services.ServerInternalKsqlClient;
+import io.confluent.ksql.services.ServiceContext;
+import java.time.Duration;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Supplier;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+@Path("/healthcheck")
+@Produces({Versions.KSQL_V1_JSON, MediaType.APPLICATION_JSON})
+public class HealthCheckResource {
+  private final HealthCheckAgent healthCheckAgent;
+  private final ResponseCache responseCache;
+
+  @VisibleForTesting
+  HealthCheckResource(
+      final HealthCheckAgent healthCheckAgent,
+      final Duration healthCheckInterval,
+      final Supplier<Long> currentTimeSupplier
+  ) {
+    this.healthCheckAgent = Objects.requireNonNull(healthCheckAgent, "healthCheckAgent");
+    this.responseCache = new ResponseCache(currentTimeSupplier, healthCheckInterval);
+  }
+
+  @GET
+  public Response checkHealth() {
+    return Response.ok(getResponse()).build();
+  }
+
+  private HealthCheckResponse getResponse() {
+    final Optional<HealthCheckResponse> response = responseCache.get();
+    if (response.isPresent()) {
+      return response.get();
+    }
+
+    final HealthCheckResponse fresh = healthCheckAgent.checkHealth();
+    responseCache.cache(fresh);
+    return fresh;
+  }
+
+  public static HealthCheckResource create(
+      final KsqlResource ksqlResource,
+      final ServiceContext serviceContext,
+      final KsqlRestConfig restConfig
+  ) {
+    return new HealthCheckResource(
+        new HealthCheckAgent(
+            new ServerInternalKsqlClient(ksqlResource, serviceContext),
+            restConfig),
+        Duration.ofMillis(restConfig.getLong(KsqlRestConfig.KSQL_HEALTHCHECK_INTERVAL_MS_CONFIG)),
+        System::currentTimeMillis
+    );
+  }
+
+  /* Caches a HealthCheckResponse for the specified duration */
+  private static class ResponseCache {
```

Comment:
Done, though the downside is I can no longer use a custom time supplier in the unit tests. One of the updated tests currently sleeps for 11 ms, but hopefully that's OK?

(Also, thanks to @agavra for pointing out that switching to Guava's implementation means we get concurrency handling for free!)

Reply:
Great!. Yes, guava will help you with concurrency as well. Regarding the sleep time, can the test use a 0ms interval health check, then sleep for 1ms in the test so you check the past health check was evicted?

Btw, I think you can even reduce more code by using the `CacheLoader` and not the `ResponseCache`. Here's a code snippet I just imagined how this would look like:
```
private static final boolean HEALTHCHECK = true;

private static final CacheLoader<Boolean, HealthCheckResponse> HEALTHCHECK_LOADER 
  = new CacheLoader<Boolean, HealthCheckResponse>() {
      @Override
      public HealthCheckResponse load(Boolean a) throws Exception {
       return healthCheckAgent.checkHealth();
      }
    };

private final LoadingCache<Boolean, HealthCheckResponse> cache;

HealthCheckResource(final Duration healthCheckInterval) {
  cache = CacheBuilder.newBuilder()
      .expireAfterWrite(healthCheckInterval.toMillis(), TimeUnit.MILLISECONDS)
      .build(HEALTHCHECK_LOADER);
}

@GET
public Response checkHealth() {
  return Response.ok(getResponse()).build();
}

private HealthCheckResponse getResponse() {
  // This will call the healthCheckAgent.checkHealth() when the interval time has passed
  return cache.getUnchecked(HEALTHCHECK);
}
```

Reply:
> Regarding the sleep time, can the test use a 0ms interval health check, then sleep for 1ms in the test so you check the past health check was evicted?

See discussion here: https://github.com/confluentinc/ksql/pull/3501#discussion_r334705444

> Btw, I think you can even reduce more code by using the `CacheLoader` and not the `ResponseCache`.

Sure, I can update to use a `LoadingCache` instead. My reason for wrapping the Guava cache in an internal `ResponseCache` class was to ensure that only a single key is ever used with the cache, but I can move this enforcement into the `CacheLoader` so I'll do that instead.